front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/handleParseFiles.ts

```diff
@@ -9,71 +9,94 @@ export const handleFileReadingError = (error: Error) => {
   console.error('File reading error:', error);
 };
 
-export const handleJsonParsingError = (error: Error, dispatch: Dispatch) => {
-  console.error('Error parsing JSON:', error);
-  dispatch(
-    setFailure({
-      name: t('errorMessages.error'),
-      message: t('errorMessages.errorInvalidJSONFormat'),
-    })
-  );
-};
+const validateTrainSchedules = (
+  importedTrainSchedules: Record<string, unknown>[]
+): TrainScheduleBase[] | null => {
+  const isInvalidTrainSchedules = importedTrainSchedules.some((trainSchedule) => {
+    if (
+      ['constraint_distribution', 'path', 'rolling_stock_name', 'start_time', 'train_name'].some(
+        (key) => !(key in trainSchedule)
+      ) ||
+      !Array.isArray(trainSchedule.path)
+    ) {
+      return true;
+    }
+    const hasInvalidSteps = trainSchedule.path.some((step) => ['id'].some((key) => !(key in step)));
+    return hasInvalidSteps;
+  });
 
-export const handleXmlParsingError = (error: Error, dispatch: Dispatch) => {
-  console.error('Error parsing XML/RailML:', error);
-  dispatch(
-    setFailure({
-      name: t('errorMessages.error'),
-      message: t('errorMessages.errorInvalidXMLFormat'),
-    })
-  );
+  if (isInvalidTrainSchedules) {
+    throw new Error('Invalid train schedules: some compulsory keys are missing');
+  }
+  return importedTrainSchedules as TrainScheduleBase[];
 };
 
 export const processJsonFile = (
   fileContent: string,
+  fileExtension: string,
   setTrainsJsonData: (data: TrainScheduleBase[]) => void,
-  dispatch: Dispatch
+  dispatch: Dispatch,
+  t: TFunction
 ) => {
+  const isJsonFile = fileExtension === 'application/json';
+
+  // try to parse the file content
+  let rawContent;
+  try {
+    rawContent = JSON.parse(fileContent);
+  } catch {
+    if (isJsonFile) {
+      dispatch(
+        setFailure({
+          name: t('errorMessages.error'),
+          message: t('errorMessages.errorInvalidFile'),
+        })
+      );
+    }
+    return isJsonFile;
+  }
+
+  // validate the trainSchedules
   try {
-    const importedTrainSchedules: TrainScheduleBase[] = JSON.parse(fileContent);
+    const importedTrainSchedules = validateTrainSchedules(rawContent);
     if (importedTrainSchedules && importedTrainSchedules.length > 0) {
```

Comment:
If I understood correctly, if we miss a compulsory key in `rawContent`, we will throw an error (in `validateTrainSchedules`) so we will fallback in `catch` here with `errorInvalidFile`. 

So same thing as before, checking `if (importedTrainSchedules &&...` isn't needed since `validateTrainSchedules` will always return an array (empty or not) if we are still in the `try`.

Reply:
Done